pkg/prometheus/agent/daemonset.go

```diff
@@ -0,0 +1,252 @@
+
+	if cpf.PortName == "" {
+		cpf.PortName = prompkg.DefaultPortName
+	}
+
+	// We need to re-set the common fields because cpf is only a copy of the original object.
+	// We set some defaults if some fields are not present, and we want those fields set in the original Prometheus object before building the DaemonSetSpec.
+	p.SetCommonPrometheusFields(cpf)
+	spec, err := makeDaemonSetSpec(p, config, cg, tlsSecrets)
+	if err != nil {
+		return nil, fmt.Errorf("make DaemonSet spec: %w", err)
+	}
+
+	// do not transfer kubectl annotations to the daemonset so it is not
+	// pruned by kubectl
+	annotations := make(map[string]string, 0)
+	for key, value := range objMeta.GetAnnotations() {
+		if !strings.HasPrefix(key, "kubectl.kubernetes.io/") {
+			annotations[key] = value
+		}
+	}
+	daemonSet := &appsv1.DaemonSet{Spec: *spec}
+
+	operator.UpdateObject(
+		daemonSet,
+		operator.WithName(name),
+		operator.WithAnnotations(annotations),
+		operator.WithAnnotations(config.Annotations),
+		operator.WithLabels(objMeta.GetLabels()),
+		operator.WithLabels(map[string]string{
+			prompkg.PrometheusNameLabelName: objMeta.GetName(),
+			prompkg.PrometheusModeLabeLName: prometheusMode,
+		}),
+		operator.WithLabels(config.Labels),
+		operator.WithManagingOwner(p),
+	)
+
+	if cpf.ImagePullSecrets != nil && len(cpf.ImagePullSecrets) > 0 {
+		daemonSet.Spec.Template.Spec.ImagePullSecrets = cpf.ImagePullSecrets
+	}
+
+	if cpf.HostNetwork {
+		daemonSet.Spec.Template.Spec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+
+	return daemonSet, nil
+}
+
+func makeDaemonSetSpec(
+	p monitoringv1.PrometheusInterface,
+	c *prompkg.Config,
+	cg *prompkg.ConfigGenerator,
+	tlsSecrets *operator.ShardedSecret,
+) (*appsv1.DaemonSetSpec, error) {
+	cpf := p.GetCommonPrometheusFields()
+
+	pImagePath, err := operator.BuildImagePathForAgent(
+		ptr.Deref(cpf.Image, ""),
+		c.PrometheusDefaultBaseImage,
+		operator.StringValOrDefault(cpf.Version, operator.DefaultPrometheusVersion),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if !slices.Contains(cpf.EnableFeatures, "agent") {
+		cpf.EnableFeatures = append(cpf.EnableFeatures, "agent")
+	}
+
+	promArgs := buildAgentArgs(cpf, cg)
+
+	volumes, promVolumeMounts, err := prompkg.BuildCommonVolumes(p, tlsSecrets)
+	if err != nil {
+		return nil, err
+	}
+
+	configReloaderVolumeMounts := prompkg.CreateConfigReloaderVolumeMounts()
+
+	var configReloaderWebConfigFile string
+
+	confArg, configVol, configMount, err := prompkg.BuildWebconfig(cpf, p)
+	if err != nil {
+		return nil, err
+	}
+
+	promArgs = append(promArgs, confArg)
+	volumes = append(volumes, configVol...)
+	promVolumeMounts = append(promVolumeMounts, configMount...)
+
+	// To avoid breaking users deploying an old version of the config-reloader image.
+	// TODO: remove the if condition after v0.72.0.
+	if cpf.Web != nil {
+		configReloaderWebConfigFile = confArg.Value
+		configReloaderVolumeMounts = append(configReloaderVolumeMounts, configMount...)
+	}
```

Comment:
@ArthurSens @simonpasquier @kakkoyun We already agreed on Prometheus version in https://github.com/prometheus-operator/prometheus-operator/pull/6652#discussion_r1653854976 What about config reloader version here? Do we need this version check?

Reply:
No, let's keep it simple.